README
NAME
    Beam::Runner - Task runner and Map-Reduce framework for Beam

VERSION
    version 0.001

SYNOPSIS
        # Job is a series of Steps
        # Steps accepts input (arguments) to become a Task
        # Task is the intersection of Step and its arguments

        # Run a job locally
        my $job = Beam::Blast::Job->new(
            wire => $wire,
            steps => [
                {
                    ref => 'foo',
                    method => 'bar',
                },
                {
                    ref => 'baz',
                    method => 'fuzz',
                    args => [], # Don't pass the output from the prior task
                },
                {
                    ref => 'fizz',
                    method => 'buzz',
                    args => [ 'moo', { '$ref': '@_' } ], # Append the output vars (curry)
                },
            ],
        );

        my ( $output_1, $output_2 ) = $job->run( $input_1, $input_2 );

        # Run using the distributed workers
        my $runner = Beam::Blast::Runner->new(
            host => 'localhost',
            port => 2132,
        );
        my $job = $runner->job(
            wire => $wire,
            steps => [
                {
                    ref => 'foo',
                    method => 'bar',
                },
                {
                    ref => 'baz',
                    method => 'fuzz',
                },
            ],
        );
        my ( $output_1, $output_2 ) = $job->run( $input_1, $input_2 );

        # Run distributed and async
        my @output;
        my $cb = sub {
            my ( $job, $output ) = @_;
            push @output, $output;
        };
        $job->run( $input_1, $input_2, $cb );

        # Run a worker
        my $worker = Beam::Blast::Worker->new(
            port => 2132,
        );
        $worker->start;

        # Run a master
        # Masters can run map-reduce jobs, parcelling out work to other workers
        # Otherwise, they appear from the outside to be workers
        my $master = Beam::Blast::Master->new(
            port => 2132,
            workers => 10,          # fork 10 child workers why not?
        );
        my $job = $master->job( ... );
        my @output = $job->run( 1..500 ); # Run all 500 tasks

        # Connect any random worker to a master
        Beam::Blast::Worker->new(
            master => 'tcp://example.com:2132',
            workers => 10,          # fork 10 more child workers
        );
        # When connected to a master, we don't need to accept any random job by
        # opening a port. But we can, and then we would transmit our status
        # to the master

        # If you want to build a status monitor, you can use the Master's event
        # callbacks

DESCRIPTION
AUTHOR
    Doug Bell <preaction@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2015 by Doug Bell.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Changelog for Beam-Runner

{{ $NEXT }}

 - Initial release



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.023.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Task runner and Map-Reduce framework for Beam",
  "dist_author" => [
    "Doug Bell <preaction\@cpan.org>"
  ],
  "dist_name" => "Beam-Runner",
  "dist_version" => "0.001",
  "license" => "perl",
  "module_name" => "Beam::Runner",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "perl" => "5.008"
  },
  "script_files" => [],
  "test_requires" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Deep" => 0,
    "Test::Differences" => 0,
    "Test::Exception" => 0,
    "Test::More" => "1.001005"
  }
);


my %fallback_build_requires = (
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.28",
  "Test::Deep" => 0,
  "Test::Differences" => 0,
  "Test::Exception" => 0,
  "Test::More" => "1.001005"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.023.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Task runner and Map-Reduce framework for Beam",
  "AUTHOR" => "Doug Bell <preaction\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.28"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "DISTNAME" => "Beam-Runner",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Beam::Runner",
  "PREREQ_PM" => {},
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Deep" => 0,
    "Test::Differences" => 0,
    "Test::Exception" => 0,
    "Test::More" => "1.001005"
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.28",
  "Test::Deep" => 0,
  "Test::Differences" => 0,
  "Test::Exception" => 0,
  "Test::More" => "1.001005"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Beam/Runner.pm
package Beam::Runner;
# ABSTRACT: Task runner and Map-Reduce framework for Beam

use strict;
use warnings;



1;
__END__

=head1 SYNOPSIS

    # Job is a series of Steps
    # Steps accepts input (arguments) to become a Task
    # Task is the intersection of Step and its arguments

    # Run a job locally
    my $job = Beam::Blast::Job->new(
        wire => $wire,
        steps => [
            {
                ref => 'foo',
                method => 'bar',
            },
            {
                ref => 'baz',
                method => 'fuzz',
                args => [], # Don't pass the output from the prior task
            },
            {
                ref => 'fizz',
                method => 'buzz',
                args => [ 'moo', { '$ref': '@_' } ], # Append the output vars (curry)
            },
        ],
    );

    my ( $output_1, $output_2 ) = $job->run( $input_1, $input_2 );

    # Run using the distributed workers
    my $runner = Beam::Blast::Runner->new(
        host => 'localhost',
        port => 2132,
    );
    my $job = $runner->job(
        wire => $wire,
        steps => [
            {
                ref => 'foo',
                method => 'bar',
            },
            {
                ref => 'baz',
                method => 'fuzz',
            },
        ],
    );
    my ( $output_1, $output_2 ) = $job->run( $input_1, $input_2 );

    # Run distributed and async
    my @output;
    my $cb = sub {
        my ( $job, $output ) = @_;
        push @output, $output;
    };
    $job->run( $input_1, $input_2, $cb );

    # Run a worker
    my $worker = Beam::Blast::Worker->new(
        port => 2132,
    );
    $worker->start;

    # Run a master
    # Masters can run map-reduce jobs, parcelling out work to other workers
    # Otherwise, they appear from the outside to be workers
    my $master = Beam::Blast::Master->new(
        port => 2132,
        workers => 10,          # fork 10 child workers why not?
    );
    my $job = $master->job( ... );
    my @output = $job->run( 1..500 ); # Run all 500 tasks

    # Connect any random worker to a master
    Beam::Blast::Worker->new(
        master => 'tcp://example.com:2132',
        workers => 10,          # fork 10 more child workers
    );
    # When connected to a master, we don't need to accept any random job by
    # opening a port. But we can, and then we would transmit our status
    # to the master

    # If you want to build a status monitor, you can use the Master's event
    # callbacks

=head1 DESCRIPTION

